crates/volta-migrate/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+//! Provides types for modeling the current state of the Volta directory and for migrating between versions
+//!
+//! A new layout should be represented by it's own struct (as in the existing v0 or v1 modules)
+//! Migrations between types should be represented by `TryFrom` implementations between the layout types
+//! (see v1.rs for examples)
+//!
+//! NOTE: Since the layout file is written once the migration is complete, all migration implementations
+//! need to be aware that they may be partially applied (if something fails in the process) and should be
+//! able to re-start gracefully from an interrupted migration
+
+#![cfg(feature = "volta-updates")]
+use std::collections::HashSet;
+use std::convert::TryInto;
+use std::fs::{DirEntry, Metadata};
+use std::path::Path;
+
+mod empty;
+mod v0;
+mod v1;
+
+use v0::V0;
+use v1::V1;
+
+use log::debug;
+use volta_core::error::ErrorDetails;
+use volta_core::fs::read_dir_eager;
+use volta_core::layout::volta_home;
+#[cfg(unix)]
+use volta_core::layout::volta_install;
+use volta_core::shim;
+use volta_fail::{Fallible, ResultExt};
+
+enum MigrationState {
```

Comment:
It'd be helpful to have a comment explaining what a `MigrationState` represents. The state of the system when a migration was requested? The state of a migration at each step throughout the migration? ...?

Reply:
Definitely will do! :+1: